File: server/schemas/api_test_case.py
from typing import Optional
from datetime import datetime
from pydantic import BaseModel


# ============================================================
# 子表 Schemas
# ============================================================

class HeaderItem(BaseModel):
    enabled: bool = True
    key: str
    value: Optional[str] = ""
    description: Optional[str] = ""
    sort_order: int = 0


class QueryParamItem(BaseModel):
    enabled: bool = True
    key: str
    value: Optional[str] = ""
    description: Optional[str] = ""
    sort_order: int = 0


class BodyFormItem(BaseModel):
    enabled: bool = True
    key: str
    value: Optional[str] = ""
    param_type: str = "text"  # text / file
    description: Optional[str] = ""
    sort_order: int = 0


class BodyRawItem(BaseModel):
    content: Optional[str] = ""


class AssertionItem(BaseModel):
    assertion_type: str  # status_code / response_time / jsonpath / header / body_contains
    operator: str  # equals / not_equals / contains / greater_than / less_than / regex / exists
    field: Optional[str] = ""
    expected: Optional[str] = ""
    description: Optional[str] = ""
    sort_order: int = 0


class ExtractItem(BaseModel):
    name: str
    source: str  # jsonpath / regex / header
    expression: str
    default_value: Optional[str] = ""
    description: Optional[str] = ""
    sort_order: int = 0


class DataRuleItem(BaseModel):
    """数据规则配置项"""
    name: str  # 变量名
    rule_type: str  # extract / static / generate / transform / conditional
    enabled: bool = True
    description: Optional[str] = ""
    default_value: Optional[str] = ""
    sort_order: int = 0

    # extract 类型专用
    source: Optional[str] = None  # jsonpath / regex / header
    expression: Optional[str] = None

    # static 类型专用
    static_value: Optional[str] = None

    # generate 类型专用
    generator: Optional[str] = None  # timestamp / uuid / random_int / random_string / now
    generator_params: Optional[dict] = None

    # transform 类型专用
    source_variable: Optional[str] = None
    transform_type: Optional[str] = None  # substring / concat / replace / upper / lower / trim / to_int / to_string / format_date
    transform_params: Optional[dict] = None

    # conditional 类型专用
    condition_variable: Optional[str] = None
    condition_operator: Optional[str] = None  # equals / not_equals / contains / is_empty / is_not_empty / greater_than / less_than
    condition_value: Optional[str] = None
    true_value: Optional[str] = None
    false_value: Optional[str] = None


class AuthConfig(BaseModel):
    auth_type: str = "none"  # none / bearer / basic / api_key
    token: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None
    api_key_name: Optional[str] = None
    api_key_value: Optional[str] = None
    api_key_location: Optional[str] = None  # header / query


# ============================================================
# 用例主表 Schemas
# ============================================================

class APITestCaseCreate(BaseModel):
    project_id: int
    environment_id: Optional[int] = None
    module: Optional[str] = None
    name: str
    description: Optional[str] = None
    preconditions: Optional[str] = None
    remark: Optional[str] = None

    method: str = "GET"
    url: str = ""
    body_type: str = "none"  # none / form-data / x-www-form-urlencoded / raw-json / raw-xml / raw-text
    auth_type: str = "none"

    setup_script: Optional[str] = None
    teardown_script: Optional[str] = None

    priority: str = "P2"  # P0 / P1 / P2 / P3
    status: str = "draft"  # draft / reviewed / deprecated

    # 嵌套子表数据
    headers: list[HeaderItem] = []
    query_params: list[QueryParamItem] = []
    body_form: list[BodyFormItem] = []
    body_raw: Optional[BodyRawItem] = None
    assertions: list[AssertionItem] = []
    extracts: list[ExtractItem] = []
    data_rules: list[DataRuleItem] = []
    auth: Optional[AuthConfig] = None


class APITestCaseUpdate(BaseModel):
    environment_id: Optional[int] = None
    module: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    preconditions: Optional[str] = None
    remark: Optional[str] = None

    method: Optional[str] = None
    url: Optional[str] = None
    body_type: Optional[str] = None
    auth_type: Optional[str] = None

    setup_script: Optional[str] = None
    teardown_script: Optional[str] = None

    priority: Optional[str] = None
    status: Optional[str] = None

    headers: Optional[list[HeaderItem]] = None
    query_params: Optional[list[QueryParamItem]] = None
    body_form: Optional[list[BodyFormItem]] = None
    body_raw: Optional[BodyRawItem] = None
    assertions: Optional[list[AssertionItem]] = None
    extracts: Optional[list[ExtractItem]] = None
    data_rules: Optional[list[DataRuleItem]] = None
    auth: Optional[AuthConfig] = None


class APITestCaseInDB(BaseModel):
    id: int
    project_id: int
    environment_id: Optional[int] = None
    environment_name: Optional[str] = None
    case_number: Optional[str] = None
    module: Optional[str] = None
    name: str
    description: Optional[str] = None
    preconditions: Optional[str] = None
    remark: Optional[str] = None

    method: str
    url: str
    body_type: str = "none"
    auth_type: str = "none"

    setup_script: Optional[str] = None
    teardown_script: Optional[str] = None

    priority: str = "P2"
    status: str = "draft"

    creator_id: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    # 嵌套子表
    headers: list[HeaderItem] = []
    query_params: list[QueryParamItem] = []
    body_form: list[BodyFormItem] = []
    body_raw: Optional[BodyRawItem] = None
    assertions: list[AssertionItem] = []
    extracts: list[ExtractItem] = []
    data_rules: list[DataRuleItem] = []
    auth: Optional[AuthConfig] = None

    class Config:
        from_attributes = True


# ============================================================
# 批量操作 Schemas
# ============================================================

class BatchDeleteRequest(BaseModel):
    case_ids: list[int]


class CreateModuleRequest(BaseModel):
    project_id: int
    module: str


class RenameModuleRequest(BaseModel):
    project_id: int
    old_module: str
    new_module: str


# ============================================================
# cURL导入 Schemas
# ============================================================

class CurlImportRequest(BaseModel):
    curl_command: str


# ============================================================
# 环境 Schemas
# ============================================================

class EnvironmentBase(BaseModel):
    name: str
    base_url: Optional[str] = None
    variables: dict = {}


class EnvironmentCreate(EnvironmentBase):
    project_id: int


class EnvironmentUpdate(BaseModel):
    name: Optional[str] = None
    base_url: Optional[str] = None
    variables: Optional[dict] = None


class EnvironmentInDB(EnvironmentBase):
    id: int
    project_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/api/apiTestCase.ts
import request from '@/utils/request'

// ==================== 子表接口 ====================

export interface HeaderItem {
  enabled: boolean
  key: string
  value: string
  description?: string
  sort_order?: number
}

export interface QueryParamItem {
  enabled: boolean
  key: string
  value: string
  description?: string
  sort_order?: number
}

export interface BodyFormItem {
  enabled: boolean
  key: string
  value: string
  param_type: 'text' | 'file'
  description?: string
  sort_order?: number
}

export interface BodyRawItem {
  content: string
}

export interface AssertionItem {
  assertion_type: 'status_code' | 'response_time' | 'jsonpath' | 'header' | 'body_contains'
  operator: 'equals' | 'not_equals' | 'contains' | 'greater_than' | 'less_than' | 'regex' | 'exists'
  field?: string
  expected?: string
  description?: string
  sort_order?: number
}

export interface DataRuleItem {
  name: string
  rule_type: 'extract' | 'static' | 'generate' | 'transform' | 'conditional'
  enabled?: boolean
  description?: string
  default_value?: string
  sort_order?: number
  // extract 类型专用
  source?: 'jsonpath' | 'regex' | 'header'
  expression?: string
  // static 类型专用
  static_value?: string
  // generate 类型专用
  generator?: 'timestamp' | 'uuid' | 'random_int' | 'random_string' | 'now'
  generator_params?: Record<string, any>
  // transform 类型专用
  source_variable?: string
  transform_type?: 'substring' | 'concat' | 'replace' | 'upper' | 'lower' | 'trim' | 'to_int' | 'to_string' | 'format_date'
  transform_params?: Record<string, any>
  // conditional 类型专用
  condition_variable?: string
  condition_operator?: 'equals' | 'not_equals' | 'contains' | 'is_empty' | 'is_not_empty' | 'greater_than' | 'less_than'
  condition_value?: string
  true_value?: string
  false_value?: string
}

export interface AuthConfig {
  auth_type: 'none' | 'bearer' | 'basic' | 'api_key'
  token?: string
  username?: string
  password?: string
  api_key_name?: string
  api_key_value?: string
  api_key_location?: 'header' | 'query'
}

// ==================== 主表接口 ====================

export interface APITestCase {
  id: number
  project_id: number
  environment_id?: number
  environment_name?: string
  case_number?: string
  module?: string
  name: string
  description?: string
  preconditions?: string
  remark?: string
  method: string
  url: string
  body_type: string
  auth_type: string
  setup_script?: string
  teardown_script?: string
  priority: string
  status: string
  creator_id?: number
  created_at: string
  updated_at?: string
  headers: HeaderItem[]
  query_params: QueryParamItem[]
  body_form: BodyFormItem[]
  body_raw?: BodyRawItem
  assertions: AssertionItem[]
  data_rules: DataRuleItem[]
  auth?: AuthConfig
}

export interface APITestCaseCreate {
  project_id: number
  environment_id?: number
  module?: string
  name: string
  description?: string
  preconditions?: string
  remark?: string
  method?: string
  url?: string
  body_type?: string
  auth_type?: string
  setup_script?: string
  teardown_script?: string
  priority?: string
  status?: string
  headers?: HeaderItem[]
  query_params?: QueryParamItem[]
  body_form?: BodyFormItem[]
  body_raw?: BodyRawItem
  assertions?: AssertionItem[]
  data_rules?: DataRuleItem[]
  auth?: AuthConfig
}

export interface APITestCaseUpdate {
  environment_id?: number
  module?: string
  name?: string
  description?: string
  preconditions?: string
  remark?: string
  method?: string
  url?: string
  body_type?: string
  auth_type?: string
  setup_script?: string
  teardown_script?: string
  priority?: string
  status?: string
  headers?: HeaderItem[]
  query_params?: QueryParamItem[]
  body_form?: BodyFormItem[]
  body_raw?: BodyRawItem
  assertions?: AssertionItem[]
  data_rules?: DataRuleItem[]
  auth?: AuthConfig
}

export interface ModuleTree {
  project_id: number
  modules: string[]
}

// ==================== cURL导入 ====================

export interface CurlParseResult {
  method: string
  url: string
  headers: HeaderItem[]
  query_params: QueryParamItem[]
  body_type: string
  body_raw_content: string
  body_form: BodyFormItem[]
  auth_type: string
  auth?: AuthConfig
}

// ==================== 环境 ====================

export interface Environment {
  id: number
  project_id: number
  name: string
  base_url?: string
  variables: Record<string, string>
  created_at: string
  updated_at?: string
}

export interface EnvironmentCreate {
  project_id: number
  name: string
  base_url?: string
  variables?: Record<string, string>
}

export interface EnvironmentUpdate {
  name?: string
  base_url?: string
  variables?: Record<string, string>
}

// ==================== 模板 ====================

export interface TestCaseTemplate {
  name: string
  description: string
  method: string
  url: string
  headers?: HeaderItem[]
  query_params?: QueryParamItem[]
  body_type?: string
  body_raw_content?: string
  assertions?: AssertionItem[]
  priority?: string
}

// ==================== API 函数 ====================

export const getTestCases = (params?: {
  skip?: number
  limit?: number
  project_id?: number
  module?: string
  keyword?: string
  priority?: string
  status?: string
}) => {
  return request<APITestCase[]>({
    url: '/api-test-cases',
    method: 'get',
    params
  })
}

export const createTestCase = (data: APITestCaseCreate) => {
  return request<APITestCase>({
    url: '/api-test-cases',
    method: 'post',
    data
  })
}

export const getTestCase = (id: number) => {
  return request<APITestCase>({
    url: `/api-test-cases/${id}`,
    method: 'get'
  })
}

export const updateTestCase = (id: number, data: APITestCaseUpdate) => {
  return request<APITestCase>({
    url: `/api-test-cases/${id}`,
    method: 'put',
    data
  })
}

export const deleteTestCase = (id: number) => {
  return request({
    url: `/api-test-cases/${id}`,
    method: 'delete'
  })
}

export const batchDelete = (case_ids: number[]) => {
  return request({
    url: '/api-test-cases/batch-delete',
    method: 'post',
    data: { case_ids }
  })
}

export const copyTestCase = (id: number) => {
  return request<APITestCase>({
    url: `/api-test-cases/${id}/copy`,
    method: 'post'
  })
}

export const getModuleTree = (project_id?: number) => {
  return request<ModuleTree[]>({
    url: '/api-test-cases/modules/tree',
    method: 'get',
    params: { project_id }
  })
}

export const createModule = (project_id: number, module: string) => {
  return request({
    url: '/api-test-cases/modules',
    method: 'post',
    data: { project_id, module }
  })
}

export const deleteModule = (project_id: number, module: string) => {
  return request({
    url: '/api-test-cases/modules',
    method: 'delete',
    params: { project_id, module }
  })
}

export const renameModule = (project_id: number, old_module: string, new_module: string) => {
  return request({
    url: '/api-test-cases/modules',
    method: 'put',
    data: { project_id, old_module, new_module }
  })
}

export const importCurl = (curl_command: string) => {
  return request<CurlParseResult>({
    url: '/api-test-cases/import-curl',
    method: 'post',
    data: { curl_command }
  })
}

export const getTemplates = () => {
  return request<TestCaseTemplate[]>({
    url: '/api-test-cases/templates',
    method: 'get'
  })
}

// ==================== 调试执行 ====================

export interface RunRequest {
  environment_id?: number
  variables?: Record<string, string>
}

export interface AssertionRunResult {
  assertion_type: string
  field?: string
  operator: string
  expected: string
  actual?: string
  passed: boolean
  error?: string
}

export interface RunResult {
  status: 'pass' | 'fail' | 'error'
  request_snapshot?: Record<string, any>
  response_info?: {
    status_code: number
    headers: Record<string, string>
    body: string
    elapsed_ms: number
    size_bytes: number
    truncated: boolean
  }
  assertions: AssertionRunResult[]
  extracted_variables: Record<string, string>
  data_rule_variables: Record<string, string>
  script_output: Record<string, any>
  error_message?: string
  duration_ms: number
}

export const runTestCase = (caseId: number, data: RunRequest) => {
  return request<RunResult>({
    url: `/test-runner/${caseId}/run`,
    method: 'post',
    data
  })
}
